### ISSUE TITLE
Automatically select "best" match if there is conflict with keywords using embedded arguments

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#4454]
Currently the following test (adapted from an example in a [comment to #4069](https://github.com/robotframework/robotframework/issues/4069#issuecomment-1237038936)) fails:

```robotframework
*** Test Cases ***
Example
    Execute "x" on device "y"

*** Keywords ***
Execute "${command}"
    Log    Executing ${command}.

Execute "${command}" on device "${device}"
    Log    Executing ${command} on ${device}.
```

The reason for the failure is that `Execute "x" on device "y"` matches both of the keyword implementations. A solution to this problem, recommended in the [User Guide](http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#using-custom-regular-expressions), has been using custom patterns that limit what keywords match. For example, changing the first keyword to `Execute "${command:[^"]+}"` would mitigate the issue.

Issue #4069 proposes making custom patterns more strict so that they would be validated also when the argument is given as a variable. As the comment referenced above explains, this unfortunately causes problems if patterns have been used for resolving conflicts as explained above.

A solution to the problem caused by conflicts, that would also allow us making custom pattern validation more strict, would be Robot somehow being able to detect what is the "best" match. For a human it is pretty clear that if you use `Execute "x" on device "y"`, `Execute "${command}" on device "${device}"` matches better than `Execute "${command}"`, but the problem is how Robot can detect that. It's also a question that can it always detect it right or could there be false positives. I have one pretty simple approach in mind that I explain in a separate comment. Other approaches can also be proposed!

**Comment by pekkaklarck** [comment]
The simplest solution I can think of, that also ought to work, is checking which of the matched keywords has longest "non argument" part. That part must match the used keyword literally, and if the part is longer then the match can also be considered better. For example, `Execute "${command}"` has 10 characters that aren't part of the arguments while `Execute "${command}" on device "${device}"` has 23. In this case it would thus be easy to tell that the latter keyword matches better. Do others see counter examples where this simple algorithm wouldn't work right?

A benefit of this solution would be that it is simple to implement. In fact I already have a prototype done and could easily finish it still in RF 5.1 timeline. I'm mostly worried if the solution good enough and how big risk there is for false positives.

**Comment by pekkaklarck** [comment]
Here's one example where the algorithm I proposed above doesn't work too well. I would say that `If / else` matches both implementations equally well and using it should cause an error. With the proposed algorithm it would match the first implementation because `${if} / else` has 7 characters that aren't part of arguments while `If / ${else}` has only 5. I need to think could this situation be detected without making the algorithm overly complicated.

```robotframework
*** Test Cases ***
Example
    x / else
    If / y
    If / else

*** Keywords ***
${if} / else
    Log to console    if: ${if}

If / ${else}
    Log to console    else: ${else}
```

**Comment by oboehmer** [comment]
thanks for looking into this, @pekkaklarck .. I performed some initial tests which look very promising, currently running our full test suite with this patch and with `[^]+` regex removed from all of our keywords.. will report status later today..

**Comment by oboehmer** [comment]
@pekkaklarck , something doesn't look right.. I do not get any more duplicate keyword errors, but the logic doesn't select the expected keyword:

```
*** Variables ***
${device}    nso_mock

*** Test Cases ***
test 1
    execute "show software | display json" on device "${device}"

test 2
    ${c}=   Set Variable    show software | display json
    execute "${c}" on device "${device}"

test 3
    execute "foo" on device "bar"

*** Keywords ***
execute "${command}"
    Fail   this should not be called

execute "${command}" on device "${device}"
    Log    this should be called
```

all tests fail:

```
robot /tmp/test.robot
==============================================================================
Test                                                                          
==============================================================================
test 1                                                                | FAIL |
this should not be called
------------------------------------------------------------------------------
test 2                                                                | FAIL |
this should not be called
------------------------------------------------------------------------------
test 3                                                                | FAIL |
this should not be called
------------------------------------------------------------------------------
Test                                                                  | FAIL |
3 tests, 0 passed, 3 failed
==============================================================================
```

**Comment by pekkaklarck** [comment]
Thanks a lot for testing @oboehmer and sorry for the stupid logic bug. Changing
```python
if other.matches(candidate.name) or not candidate.matches(other.name):
```
to
```python
if candidate.matches(other.name) or not other.matches(candidate.name):
```
makes the above tests succeed and hopefully works fine also otherwise.

This is what you get when you don't have proper automated tests. I just had some example tests and I validated their results manually. Apparently I didn't look at what they actually did after I changed the implementation. Need to create tests that validate that results are as expected.

**Comment by oboehmer** [comment]
@pekkaklarck , I wonder if we can also resolve this use case:

```
$ cat test3.robot 
*** Settings ***
Resource   res1.robot
Resource   res2.robot

*** Variables ***
${device}    foo

*** Test Cases ***
test
    disconnect from device "${device}" with alias "test_connection"

$ cat res1.robot 
# res1.robot
*** Keywords ***
disconnect from device "${name}"
    Fail   this should not be executed
$   
$ cat res2.robot 
# res2.robot
*** Keywords ***
disconnect from device "${device}" with alias "${alias}"
    Log    this should be executed
$ 
```
executing the test suite results in a `Multiple keywords` warning, albeit a very interesting one as it doesn't show the right keyword for res1:

```
Multiple keywords with name 'disconnect from device "${device}" with alias "test_connection"' found. Give the full name of the keyword you want to use:
    res1.disconnect from device "${device}" with alias "test_connection"
    res2.disconnect from device "${device}" with alias "test_connection"
```

if I put in a search order to prefer res1, the test picks the keyword in res1, which is not the right one:

```
$ cat test3.robot 
*** Settings ***
Resource   res1.robot
Resource   res2.robot
Suite Setup     Set Library Search Order    res1

*** Variables ***
${device}    foo

*** Test Cases ***
test
    disconnect from device "${device}" with alias "test_connection"

$ robot test3.robot 
==============================================================================
Test3                                                                         
==============================================================================
test                                                                  | FAIL |
this should not be executed
------------------------------------------------------------------------------
Test3                                                                 | FAIL |
1 test, 0 passed, 1 failed
==============================================================================
```

**Comment by pekkaklarck** [comment]
Conflicts with embedded args are at the moment only looked at in a single file. I agree it would be better to handle that also if kws are in different files. Need to check how easy/hard that would be.

The case with search order is a bit different. If you explicitly give precedence to s certain lib/resource and it contains a matching keyword, I think using it without looking others makes sense. How is that currently handled without patches?

Both of these cases can also be resolved by prefixing the keyword with the lib/resource name.

**Comment by oboehmer** [comment]
Well, I would expect the logic to work across libs/files, so the conflict which would require search order (or prefix) would not occur at all.. will check how this case is handled without the patch.. reckon it would be the same..

**Comment by pekkaklarck** [comment]
Conflicts across files would need to be handled here:
https://github.com/robotframework/robotframework/blob/c87a71579f5b48251006d52ec33bfaa5706d7806/src/robot/running/namespace.py#L341

Probably doable but I don't have time to look at it right now. The search order is looked at there as well, so if this conflict is handled first then better match wins even if the search order is set. I'm still not 100% is that a good idea, though.

**Comment by pekkaklarck** [comment]
If you want to test this with #4069 reverted (or disabled to be more precise), you can apply this trivial diff:
```diff
diff --git a/src/robot/running/arguments/embedded.py b/src/robot/running/arguments/embedded.py
index 9c7b2f6cf..e37c9e509 100644
--- a/src/robot/running/arguments/embedded.py
+++ b/src/robot/running/arguments/embedded.py
@@ -39,6 +39,7 @@ class EmbeddedArguments:
         return list(zip(self.args, values))
 
     def validate(self, values):
+        return
         if not self.custom_patterns:
             return
         for arg, value in zip(self.args, values):
```

I'll do it like that myself but I'll add a comment there explaining it as well. I have that already done, but I also need to update tests and docs before committing it. No time for that now.

**Comment by pekkaklarck** [comment]
Commit 9419f614dfcbfa37ae4397816d01ff8652e1488a changes keyword conflict resolution so that all conflicts are resolved on namespace. Earlier conflicts involving embedded arguments were resolved within the file where they were defined (`HandlerStore`) and others on namespace. This allows resolving conflicts involving embedded args across all files. Based on my tests (nothing committed yet) the same algorithm I used earlier works well in that case as well.

**Comment by pekkaklarck** [comment]
The main open question is deciding in which order different conflict resolution strategies should be tried out. This is especially important when conflicts occur with resource files because with them there are various such strategies. The current order is (see [code](https://github.com/robotframework/robotframework/blob/9419f614dfcbfa37ae4397816d01ff8652e1488a/src/robot/running/namespace.py#L343) is this:

1. Search order
2. Keywords in same file as the caller (new in RF 5.1, see #4366)
3. Public keywords over private keywords (new in RF 5.1, see #430)

We need to decide when do we try selecting the best embedded match, but we can also think is the above order good in general. In earlier comments @oboehmer proposed that selecting the best embedded match should happen before looking at the search order. Back then I was thinking the search order should have the highest priority, but after thinking this more I tend to agree with @oboehmer and see that the search order should be used as the last resolution mechanism if everything else fails. If we agree on that, we need to add tests for keywords in same file and public/private keywords in combination with the search order. The resulting order would then be this:

1. Keywords in same file as the caller
2. Public keywords over private keywords
3. Search order

The order of 1. and 2. cannot be changed because we want to use the local keyword over keywords elsewhere regardless is the local keyword private or not. I also believe they should be done right after each others. The open question then is, should we look at the best embedded match before them or after them? Do you @oboehmer have opinions related to that?

**Comment by pekkaklarck** [comment]
If we first look at the best embedded match, we may end up selecting a private keyword from some other file even though there would be a public matching keyword somewhere, possibly in the same file, as well. That doesn't sound too good to me, so I believe we should use this order:

1. Keywords in same file as the caller
2. Public keywords over private keywords
3. Best embedded match
4. Search order

Do others see problems with that? Most importantly, do you see cases where this order causes problems and some other order would work better?

**Comment by pekkaklarck** [comment]
The commit above changes conflict resolution order to this:

1. Keywords in same file as the caller
2. Public keywords over private keywords
3. Search order

See comments above for motivation. Still need to decide what's the correct place to look for possible best embedded match.

**Comment by oboehmer** [comment]
Thanks so much for enhancing this so quickly, didn’t expect it after reverting #4069. At first glance I would agree with the order proposed in https://github.com/robotframework/robotframework/issues/4454#issuecomment-1248038078, but give me a few days, will reply after the weekend..

**Comment by pekkaklarck** [comment]
Selecting the best match was implemented in 95881a2ae0341b846e8df9be1b7887a4d78408be. The conflict resolution order is the one discussed above.

I'm pretty happy how this works now. Docs still need to be updated and I'd also like to hear feedback from @oboehmer about the implementation before closing this.

**Comment by pekkaklarck** [comment]
Huge thanks to @oboehmer for testing this and reporting a problem related to search order. That ought to be fixed by the commit above.

**Comment by oboehmer** [comment]
It looks the latest commit now allows us to drop all of the `[^"]+` regexps from our keywords, which are scattered around multiple libraries, partially with keywords overlapping across libraries. This really allows us to prepare for the  #4069 enforcement once, and I trust we can now safely introduce it in 5.2 or 6.0, whenever..It would be good to announce this new matching capability quite loudly in the release notes so others can also prepare their libs accordingly..

**Comment by pekkaklarck** [comment]
I set high priority to this issue exactly to get it mentioned in the release notes. This enhancement makes using keywords with embedded arguments quite a bit easier regardless would #4069 affect you or not. Still need to update examples in the User Guide accordingly!

**Comment by pekkaklarck** [comment]
The section about embedded arguments in the User Guide was pretty bad in general and I ended up rewriting it pretty much totally. Great to have this issue closed now!

**Comment by pekkaklarck** [comment]
As explained in #4609, It turned out that using the search order only as the last resort wasn't a good idea. In RF 6.1 it will be used first.
